### ISSUE TITLE
Info on minimal wyrmgus version

### ISSUE THREAD

**Issue by AMDmi3** [Andrettin/Wyrmsun#14]
Currently there's no information on which minimal version of wyrmgus is required to run each specific release of wyrmsun. Maybe you could tag wyrmgus with the same versions as wyrmsun?

**Comment by akien-mga** [comment]
From my experience, you should always use the last commit before the wyrmsun tag; there were many occurrences where a new feature in the game repo requires a code change in the engine too.

Better than tagging versions in wyrmgus, I think a submodule could be added to the wymsun repo, always pointing to the relevant commit hash as both repos progress. I have little experience with submodules though, I don't know if it's possible to set them up to always track the HEAD of the other repo (still picking a precise commit when making a tag of course), or if they should always be updated manually (which could be a bother for @Andrettin).

**Comment by AMDmi3** [comment]
> From my experience, you should always use the last commit before the wyrmsun tag

Yes, I'm currently doing exactly that, but I fear that latest commits may contain bugs while not really being required for the latest game.

> I don't know if it's possible to set them up to always track the HEAD

It's not. It'd be broken and unsafe that way.

> or if they should always be updated manually (which could be a bother for @Andrettin)

I don't see it as a bother - it's as easy as git checkout (in submodule, to update to newer needed revision) + git commit (in parent repo, to say that submodule now points to another commit), however this is a most correct way of defining a requirement for specific engine commit, and it works not only for tags but for each commit.

I think this was already suggested in some another issue (either here or in wyrmgus repo) along with root CMakeLists.txt which recurse to engine submodule directory.

**Comment by akien-mga** [comment]
>> From my experience, you should always use the last commit before the wyrmsun tag

> Yes, I'm currently doing exactly that, but I fear that latest commits may contain bugs while not really being required for the latest game.

I meant the last commit which was made before the tag was created, so you have to cross-check the dates. From what I know of @Andrettin's workflow, he always develops wyrmgus and wyrmsun in parallel; I don't think he pushes potentially buggy stuff to wyrmgus, only what is needed for the latest commit in wyrmgus.

>>I don't know if it's possible to set them up to always track the HEAD

> It's not. It'd be broken and unsafe that way.

Not at all, since that's precisely how @Andrettin works. When a version is tagged in the wyrmsun repo, it was tested against the HEAD of wyrmgus. So if the submodule tracks HEAD, it should put the correct reference in the tag (I would expect the reference in the tag to be a reference to the current HEAD commit, and not to an ever-changing HEAD of course, else that would be a weird design; tags should not have dynamical content).

**Comment by AMDmi3** [comment]
> I meant the last commit which was made before the tag was created, so you have to cross-check the dates

Yes, but it's the same.

> So if the submodule tracks HEAD, it should put the correct reference in the tag (I would expect the reference in the tag to be a reference to the current HEAD commit, and not to an ever-changing HEAD of course, else that would be a weird design; tags should not have dynamical content).

I don't quite understand what you mean under this tracking HEAD without tracking HEAD, but in fact submodule always refers to a specific fixed commit of another repo, anything else would be, as said, broken and unsafe. But yes, you may update it to another commit, including current HEAD of another repo, at any time and fix that with a parent repo commit.

The point is that as soon as you need a new features from engine, you update the subrepository to get these new features, then develop and test the game with it. Thus you have the game always linked to sufficient and working version of the engine, while you can still develop the engine independently.

**Comment by Andrettin** [comment]
> Currently there's no information on which minimal version of wyrmgus is required to run each specific release of wyrmsun. Maybe you could tag wyrmgus with the same versions as wyrmsun?

Good idea. I did that now for the latest release.

> From my experience, you should always use the last commit before the wyrmsun tag; there were many occurrences where a new feature in the game repo requires a code change in the engine too.

Yes, that is exactly the version I use when compiling the windows executable :)

> Better than tagging versions in wyrmgus, I think a submodule could be added to the wymsun repo, always pointing to the relevant commit hash as both repos progress. I have little experience with submodules though, I don't know if it's possible to set them up to always track the HEAD of the other repo (still picking a precise commit when making a tag of course), or if they should always be updated manually (which could be a bother for @Andrettin).

Unfortunately, I don't have much experience with submodules either... it might be simpler to simply tag each version of Wyrmgus with the same version number as Wyrmsun. Making it automatically track the head would be very practical, though.

> I meant the last commit which was made before the tag was created, so you have to cross-check the dates. From what I know of @Andrettin's workflow, he always develops wyrmgus and wyrmsun in parallel; I don't think he pushes potentially buggy stuff to wyrmgus, only what is needed for the latest commit in wyrmgus.

Yes; there is of course always the chance that there is a bug somewhere, but I only commit things to Wyrmsun or Wyrmgus after testing them.

**Comment by akien-mga** [comment]
> I don't quite understand what you mean under this tracking HEAD without tracking HEAD, but in fact submodule always refers to a specific fixed commit of another repo, anything else would be, as said, broken and unsafe. But yes, you may update it to another commit, including current HEAD of another repo, at any time and fix that with a parent repo commit.

What I mean is that for this particular workflow, it's not desirable to have a submodule which refers to a specific commit of another repo, as the HEAD of wyrmsun will almost always need to refer to the HEAD of wyrmgus. So then if submodules can't be configured to track HEAD automatically, @Andrettin would have to do a checkout and a commit each time a change is made in wyrmgus, and this would greatly pollute the git log.

> The point is that as soon as you need a new features from engine, you update the subrepository to get these new features, then develop and test the game with it. Thus you have the game always linked to sufficient and working version of the engine, while you can still develop the engine independently.

I understand the idea; it's just not the kind of workflow that I've seen applied here. I think @Andrettin pushes mostly tested and stable features on both repos at the same time, I've rarely seen a WIP feature in wyrmgus alone.

**Comment by Andrettin** [comment]
> I understand the idea; it's just not the kind of workflow that I've seen applied here. I think @Andrettin pushes mostly tested and stable features on both repos at the same time, I've rarely seen a WIP feature in wyrmgus alone.

Yes, you are right. Almost all of my changes are small, self-contained changes which don't break anything. I am a real big fan of such an iterative approach :)

And for bigger changes I only commit them after they're finished. The only exception I can recall is the new Earth map I'm working on, which I'm committing bit by bit.

**Comment by AMDmi3** [comment]
> So then if submodules can't be configured to track HEAD automatically, @Andrettin would have to do a checkout and a commit each time a change is made in wyrmgus, and this would greatly pollute the git log.

It will not affect the log at all. You just commit submodule updates along with game changes which require them, that's all.

Anyway, there're now tags for wyrmgus so I'm satisfied.
